File: src/main/java/net/thedragonteam/thedragonlib/properties/PropertyString.java
package net.thedragonteam.thedragonlib.properties;

import com.google.common.base.Optional;
import com.google.common.collect.ImmutableSet;
import net.minecraft.block.properties.PropertyHelper;
import net.thedragonteam.thedragonlib.util.ArrayUtils;
import net.thedragonteam.thedragonlib.util.LogHelper;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public class PropertyString extends PropertyHelper<String> {

    private final List<String> valuesSet;

    public PropertyString(String name, String... values) {
        super(name, String.class);
        valuesSet = new ArrayList<>();
        Collections.addAll(valuesSet, ArrayUtils.arrayToLowercase(values));
    }

    @Nonnull
    @Override
    public Collection<String> getAllowedValues() {
        return ImmutableSet.copyOf(valuesSet);
    }

    @Nonnull
    @Override
    public Optional<String> parseValue(@Nonnull String value) {
        return valuesSet.contains(value) ? Optional.of(value) : Optional.absent();
    }

    @Nonnull
    @Override
    public String getName(@Nonnull String value) {
        return value;
    }

    public int toMeta(String value) {
        return valuesSet.contains(value) ? valuesSet.indexOf(value) : 0;
    }

    public String fromMeta(int meta) {
        if (meta >= 0 && meta < valuesSet.size()) {
            return valuesSet.get(meta);
        }
        LogHelper.error("[PropertyString] Attempted to load property for invalid meta value");
        return valuesSet.get(0);
    }
}

File: src/main/java/net/thedragonteam/thedragonlib/properties/PropertyString.kt
package net.thedragonteam.thedragonlib.properties

import com.google.common.base.Optional
import com.google.common.collect.ImmutableSet
import net.minecraft.block.properties.PropertyHelper
import net.thedragonteam.thedragonlib.util.ArrayUtils
import net.thedragonteam.thedragonlib.util.LogHelper
import java.util.*

class PropertyString(name: String, vararg values: String) : PropertyHelper<String>(name, String::class.java) {

    private val valuesSet: List<String>

    init {
        valuesSet = ArrayList<String>()
        Collections.addAll<String>(valuesSet, *ArrayUtils.arrayToLowercase(values))
    }

    override fun getAllowedValues(): Collection<String> {
        return ImmutableSet.copyOf(valuesSet)
    }

    override fun parseValue(value: String): Optional<String> {
        return if (valuesSet.contains(value)) Optional.of(value) else Optional.absent<String>()
    }

    override fun getName(value: String): String {
        return value
    }

    fun toMeta(value: String): Int {
        return if (valuesSet.contains(value)) valuesSet.indexOf(value) else 0
    }

    fun fromMeta(meta: Int): String {
        if (meta >= 0 && meta < valuesSet.size) {
            return valuesSet[meta]
        }
        LogHelper.error("[PropertyString] Attempted to load property for invalid meta value")
        return valuesSet[0]
    }
}